File: src/lib/workspaceIndex.ts
import { invoke } from "@tauri-apps/api/core";

export type DocumentType =
  | "asciiDoc"
  | "markdown"
  | "json"
  | "yaml"
  | "text"
  | "plantUml"
  | "mermaid";

export type Severity = "error" | "warning";

export type DiagnosticKind =
  | "missingInclude"
  | "missingXrefDocument"
  | "missingXrefAnchor"
  | "missingImage"
  | "duplicateAnchor"
  | "circularInclude"
  | "parseError"
  /** Атрибуты шапки отделены от заголовка пустой строкой — `:toc:` не действует. */
  | "detachedHeaderAttributes";

export type Document = {
  id: string;
  absolutePath: string;
  relativePath: string;
  fileName: string;
  docType: DocumentType;
  modifiedAt: number;
};

export type Anchor = {
  id: string;
  document: string;
  line: number;
  column: number;
};

export type Include = {
  path: string;
  sourceDocument: string;
  line: number;
  column: number;
};

export type Reference = {
  targetDocument: string;
  anchor: string | null;
  sourceDocument: string;
  line: number;
  column: number;
};

export type Attribute = {
  name: string;
  value: string;
  document: string;
  line: number;
};

export type Image = {
  path: string;
  document: string;
  line: number;
};

export type Diagnostic = {
  kind: DiagnosticKind;
  message: string;
  document: string;
  line: number;
  column: number;
  severity: Severity;
};

export type IndexStats = {
  documents: number;
  anchors: number;
  includes: number;
  references: number;
  attributes: number;
  images: number;
  warnings: number;
  errors: number;
};

export type IndexEvent =
  | { kind: "indexBuildingStarted" }
  | {
      kind: "indexBuildingProgress";
      payload: { done: number; total: number; current: string };
    }
  | { kind: "indexBuildingFinished"; payload: { stats: IndexStats } }
  | { kind: "indexUpdated"; payload: { document: string } }
  | { kind: "diagnosticsUpdated"; payload: { document: string } };

export const INDEX_EVENT_CHANNEL = "workspace-index://event";

export function buildIndex(repoRoot: string): Promise<IndexStats> {
  return invoke<IndexStats>("build_index", { repoRoot });
}

export function clearIndex(): Promise<void> {
  return invoke<void>("clear_index");
}

export function indexIsOpen(): Promise<boolean> {
  return invoke<boolean>("index_is_open");
}

export function getDocument(path: string): Promise<Document | null> {
  return invoke<Document | null>("get_document", { path });
}

/** Same lookup as `getDocument`, but by repo-relative `DocumentId` key
 * rather than a filesystem path — the backend's `get_document` resolves
 * `path` against the real filesystem (canonicalize), so it needs an
 * absolute path; this variant looks up the index's own string key
 * directly, which is all the frontend ever has on hand for a document it
 * hasn't opened yet (e.g. resolving an `include::`/`xref:` target). */
export function getDocumentById(id: string): Promise<Document | null> {
  return invoke<Document | null>("get_document_by_id", { id });
}

export function getDocuments(): Promise<Document[]> {
  return invoke<Document[]>("get_documents");
}

export function findDocument(name: string): Promise<Document[]> {
  return invoke<Document[]>("find_document", { name });
}

export function findAnchor(id: string): Promise<Anchor[]> {
  return invoke<Anchor[]>("find_anchor", { id });
}

export function findAnchors(documentId: string): Promise<Anchor[]> {
  return invoke<Anchor[]>("find_anchors", { documentId });
}

export function findIncludes(documentId: string): Promise<Include[]> {
  return invoke<Include[]>("find_includes", { documentId });
}

export function findReferences(documentId: string): Promise<Reference[]> {
  return invoke<Reference[]>("find_references", { documentId });
}

export function findAttribute(name: string): Promise<Attribute[]> {
  return invoke<Attribute[]>("find_attribute", { name });
}

export function getAttributes(documentId: string): Promise<Attribute[]> {
  return invoke<Attribute[]>("get_attributes", { documentId });
}

export function findImage(path: string): Promise<Image[]> {
  return invoke<Image[]>("find_image", { path });
}

export function getDiagnostics(): Promise<Diagnostic[]> {
  return invoke<Diagnostic[]>("get_diagnostics");
}

export function getDiagnosticsFor(documentId: string): Promise<Diagnostic[]> {
  return invoke<Diagnostic[]>("get_diagnostics_for", { documentId });
}

File: src-tauri/src/commands/workspace_index.rs
use std::path::PathBuf;
use std::sync::Arc;

use tauri::State;

use crate::domain::workspace_index::{
    Anchor, Attribute, Diagnostic, Document, Image, Include, IndexStats, Reference,
};
use crate::services::workspace_index::WorkspaceIndex;

#[tauri::command]
pub async fn build_index(
    index: State<'_, Arc<WorkspaceIndex>>,
    repo_root: String,
) -> Result<IndexStats, String> {
    let index = index.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        index
            .build(PathBuf::from(repo_root))
            .and_then(|stats| {
                index.start_watcher()?;
                Ok(stats)
            })
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub fn clear_index(index: State<'_, Arc<WorkspaceIndex>>) -> Result<(), String> {
    index.clear();
    Ok(())
}

#[tauri::command]
pub fn index_is_open(index: State<'_, Arc<WorkspaceIndex>>) -> bool {
    index.is_open()
}

#[tauri::command]
pub fn get_document(
    index: State<'_, Arc<WorkspaceIndex>>,
    path: String,
) -> Result<Option<Document>, String> {
    Ok(index.get_document(std::path::Path::new(&path)))
}

#[tauri::command]
pub fn get_documents(index: State<'_, Arc<WorkspaceIndex>>) -> Result<Vec<Document>, String> {
    Ok(index.get_documents())
}

#[tauri::command]
pub fn find_document(
    index: State<'_, Arc<WorkspaceIndex>>,
    name: String,
) -> Result<Vec<Document>, String> {
    Ok(index.find_document(&name))
}

#[tauri::command]
pub fn find_anchor(
    index: State<'_, Arc<WorkspaceIndex>>,
    id: String,
) -> Result<Vec<Anchor>, String> {
    Ok(index.find_anchor(&id))
}

#[tauri::command]
pub fn find_anchors(
    index: State<'_, Arc<WorkspaceIndex>>,
    document_id: String,
) -> Result<Vec<Anchor>, String> {
    Ok(index.find_anchors(&crate::domain::workspace_index::DocumentId::new(
        document_id,
    )))
}

#[tauri::command]
pub fn find_includes(
    index: State<'_, Arc<WorkspaceIndex>>,
    document_id: String,
) -> Result<Vec<Include>, String> {
    Ok(index.find_includes(&crate::domain::workspace_index::DocumentId::new(
        document_id,
    )))
}

#[tauri::command]
pub fn find_references(
    index: State<'_, Arc<WorkspaceIndex>>,
    document_id: String,
) -> Result<Vec<Reference>, String> {
    Ok(index.find_references(&crate::domain::workspace_index::DocumentId::new(
        document_id,
    )))
}

#[tauri::command]
pub fn find_attribute(
    index: State<'_, Arc<WorkspaceIndex>>,
    name: String,
) -> Result<Vec<Attribute>, String> {
    Ok(index.find_attribute(&name))
}

#[tauri::command]
pub fn get_attributes(
    index: State<'_, Arc<WorkspaceIndex>>,
    document_id: String,
) -> Result<Vec<Attribute>, String> {
    Ok(index.get_attributes(&crate::domain::workspace_index::DocumentId::new(
        document_id,
    )))
}

#[tauri::command]
pub fn find_image(
    index: State<'_, Arc<WorkspaceIndex>>,
    path: String,
) -> Result<Vec<Image>, String> {
    Ok(index.find_image(&path))
}

#[tauri::command]
pub fn get_diagnostics(
    index: State<'_, Arc<WorkspaceIndex>>,
) -> Result<Vec<Diagnostic>, String> {
    Ok(index.get_diagnostics())
}

#[tauri::command]
pub fn get_diagnostics_for(
    index: State<'_, Arc<WorkspaceIndex>>,
    document_id: String,
) -> Result<Vec<Diagnostic>, String> {
    Ok(index.get_diagnostics_for(&crate::domain::workspace_index::DocumentId::new(
        document_id,
    )))
}
